mgr/dashboard: Fix NFS routing

afreen23 opened the pull request:
Fixes https://tracker.ceph.com/issues/65310

The NFS tab in object and File nav uses same route due to which both
gets activated when one of them is clicked.
Hence, this PR separates the routing for Object and File nav.
Object-> NFS: /rgw/nfs
File-> NFS: /cephfs/nfs

Both routes use same NFS List and Form component but under different
routes as mentioned above.

Changes summary
- updated route for File from "/fs" to "/cephfs/<any_other_sub_route>"
  to support both fs and nfs tabs. Since using `/fs` and `/fs/nfs` will
activate both paths and it will be an undesirable user experience.
- `getFsalRouteFromPath` helper function to set the storage backend from
  route.
- removed `stoarge-backend` field from nfs form as now route decides teh
  storage backend
- breadcrumbs redirect to respective navs
- updated e2e tests
- updated unit tests
- changes list page of object-> nfs page to say Bucket instead of Path

Signed-off-by: Afreen <afreen23.git@gmail.com>

[Screencast from 2024-06-12 19-47-33.webm](https://github.com/ceph/ceph/assets/25664409/90fbc110-acc0-4544-a60e-3ecab9e3ee27)

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

afreen23 commented:
jenkins test make check arm64

afreen23 commented:
Updates:
- fixed storage backend to show cephfs and Object gateway
- fixed e2e tests
- fixed unit tests for nfs list

Working on:
- fixing unit tests for nfs form

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/app-routing.module.ts:
```diff
@@ -370,18 +370,48 @@ const routes: Routes = [
-          { path: '', component: CephfsListComponent },
           {
-            path: URLVerbs.CREATE,
+            path: 'fs',
```
why do we need to hardcode `fs` here? this will make the route `cephfs/fs` isn't it which looks a bit odd.

afreen23 replied:
Both filesystems and NFS nav will be  matched, which was not looking good from UX perspective with `/cephfs` and `/cephfs/nfs`. May be we can check for full path, but:
This approach aligns with the rest fo dashboard:
cephfs:
- /fs
-/nfs

rgw:
-/bucket
-/overview
-/nfs

nizamial09 commented on src/pybind/mgr/dashboard/frontend/cypress/e2e/orchestrator/workflow/nfs/nfs-export.po.ts:
```diff
@@ -3,21 +3,20 @@ import { PageHelper } from '../../../page-helper.po';
   pages = pages;
 
-  @PageHelper.restrictTo(pages.create.url)
+  @PageHelper.restrictTo(pages.cephfs_create.url)
```
this view could be from either `rgw` or `cephfs`. So we can't really restrict it to a particular url. I think we only have e2es written for rgw exports so right now atleast the test should start from the `/rgw/nfs` form. We can remove this `restrictTo` from here to simplify it.

afreen23 replied:
I guess I missed this one to change

afreen23 replied:
Removed now!

nizamial09 commented on src/pybind/mgr/dashboard/frontend/cypress/e2e/orchestrator/workflow/nfs/nfs-export.po.ts:
```diff
@@ -38,8 +37,8 @@ export class NFSPageHelper extends PageHelper {
-  editExport(pseudo: string, editPseudo: string) {
-    this.navigateEdit(pseudo);
+  editExport(pseudo: string, editPseudo: string, url: string) {
+    this.navigateEdit(pseudo, true, true, url);
```
or we can select and Edit the export which we want to edit manually rather than going to the edit route. But if the current one works and doesn't break any other tests, then it should be fine.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html:
```diff
@@ -52,24 +52,13 @@
                    for="name"
                    i18n>Storage Backend</label>
             <div class="cd-col-form-input">
-              <select class="form-select"
-                      formControlName="name"
-                      name="name"
-                      id="name"
-                      (change)="fsalChangeHandler()">
-                <option *ngIf="allFsals === null"
-                        value=""
-                        i18n>Loading...</option>
-                <option *ngIf="allFsals !== null && allFsals.length === 0"
-                        value=""
-                        i18n>-- No data pools available --</option>
-                <option *ngIf="allFsals !== null && allFsals.length > 0"
-                        value=""
-                        i18n>-- Select the storage backend --</option>
-                <option *ngFor="let fsal of allFsals"
-                        [value]="fsal.value"
-                        [disabled]="fsal.disabled">{{ fsal.descr }}</option>
-              </select>
```
why are we removing this?

nizamial09 replied:
if this is removing the storage back-end selector completely, instead of that when routing through cephfs, we can just select the `cephfs` as storage back-end and if rgw then the RGW can be selected?

afreen23 replied:
If user clicked nfs under Object, it should be always Object and he should not select cephfs from object.
I have changed it to readonly input therefore

afreen23 replied:
I ahve completely removed this section as its redundant.

nizamial09 reviewed (commented):
i haven't tested this manually. I'll be less confused when I try it later on.

afreen23 commented:
jenkins test dashboard

afreen23 commented:
jenkins test make check

cloudbehl commented:
@afreen23 can you check the test failures?

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/models/nfs.fsal.ts:
```diff
@@ -1,5 +1,9 @@
+export enum SUPPORTED_FSAL {
+  CEPH = 'CEPH',
```
it should ideally be FS or CEPHFS.

CEPH FSAL doesn't make sense. 

I know this was introduced earlier and not in PR but it looks confusing.

afreen23 replied:
I think this is the way its defined in backend as well https://docs.ceph.com/en/latest/cephfs/nfs/#configuring-nfs-ganesha-to-export-cephfs thats why we have it like that.

afreen23 commented:
jenkins test dashboard

cloudbehl reviewed (approved):
LGTM

afreen23 commented:
jenkins test dashboard cephadm

afreen23 commented:
Working on fixing cephadm tests

afreen23 commented:
> Working on fixing cephadm tests

Fixed now 🎈

afreen23 commented:
Summary of changes:

* fixed minor bugs in form like - bucket validations not working
* code refactors
* tested e2e 

Screencast is updated cc @cloudbehl

afreen23 commented:
Summary of changes:

* fixed bug: need to click twice when creating ceph fs based nfs export
* changed the naming for Object-> NFS list to `Bucket` instead of path, thats the way we do in nfs form previously too. And now views are separated we can call here too `Bucket`
![image](https://github.com/ceph/ceph/assets/25664409/5d2299e5-9755-482f-bc8d-f18aa63e2e80)

Updated screencast again.

afreen23 commented:
dashboard tests fixed here https://github.com/ceph/ceph/pull/57992

cloudbehl commented:
jenkins test dashboard

cloudbehl commented:
Tested this and it looks good!

nizamial09 commented:
jenkins test dashboard cephadm

afreen23 commented:
Fixed cephadm tests

afreen23 commented:
jenkins test make check arm64

afreen23 commented:
jenkins test make check

afreen23 commented:
jenkins test dashboard

afreen23 commented:
jenkins test dashboard cephadm

afreen23 commented:
jenkins test make check

afreen23 commented:
jenkins test ceph dashboard

afreen23 commented:
jenkins test ceph dashboard cephadm

afreen23 commented:
jenkins test ceph dashboard cephadm

afreen23 commented:
jenkins test dashboard cephadm

afreen23 commented:
jenkins test make check

afreen23 commented:
jenkins test dashboard

afreen23 commented:
jenkins test make check

afreen23 commented:
jenkins test dashboard

afreen23 commented:
jenkins test dashboard cephadm

afreen23 commented:
jenkins test make check arm64

afreen23 commented:
jenkins test make check arm64
